### PR TITLE
thelounge: fix build

### LINKED ISSUE #233630: Build failure: thelounge
### Steps To Reproduce

```
$ nix build --no-link -f https://github.com/NixOS/nixpkgs/archive/refs/heads/nixos-23.05.tar.gz thelounge
```

### Build log

```
linking bin 'thelounge'
can't find file to patch at input line 5
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/src/helper.js b/src/helper.js
|index 27352b53..7078e4c5 100644
|--- a/src/helper.js
|+++ b/src/helper.js
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
/nix/store/08wcfajdgcs02ivzwsl0bgg2x5jjf0vp-stdenv-linux/setup: line 136: pop_var_context: head of shell_variables not a function context
```

### Additional context

The failure above applies to both `x86_64-linux` and `aarch64-linux` builders.

### Notify maintainers

@winterqt

### Metadata

```console
 - system: `"aarch64-linux"`
 - host os: `Linux 6.3.1-zen1, NixOS, 22.11 (Raccoon), 22.11.4286.7dc71aef32e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - channels(root): `"nixos-23.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by winterqt**
###### Description of changes

Upstream switched to using TypeScript in v4.4.0, which broke the patch. This fixes that issue by migrating to building The Lounge from source, instead of having to patch the minified JavaScript.

Closes #233630.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/irc/thelounge/default.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  configurePhase = ''
+    runHook preConfigure
+
+    export HOME="$PWD"
+
+    fixup_yarn_lock yarn.lock
+    yarn config --offline set yarn-offline-mirror ${finalAttrs.offlineCache}
+    yarn install --offline --frozen-lockfile --ignore-platform --ignore-scripts --no-progress --non-interactive
+    patchShebangs node_modules
```
Comment (lilyinstarlight) [inline pkgs/applications/networking/irc/thelounge/default.nix:37]: buildYarnPackage hooks when

**Review by lilyinstarlight — COMMENTED** [review]
Assuming ofborg shows all green, this looks good to me (really want those yarn hooks though and my PR to add a fixup-yarn-lock command independent of yarn2nix...)

File: pkgs/applications/networking/irc/thelounge/default.nix (hunk 1 above)
Comment (winterqt) [reply pkgs/applications/networking/irc/thelounge/default.nix:37]: Something something #210814

**Review by JulienMalka — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
I'm going to be a guinea pig for that, pushing in production.

**Comment by RaitoBezarius** [comment]
Production is running with it, without any hiccups. :)

File: pkgs/applications/networking/irc/thelounge/default.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  # Takes way, way, way too long.
+  dontStrip = true;
+
+  passthru.tests = nixosTests.thelounge;
+
+  meta = with lib; {
+    description = "A glossy Matrix collaboration client for the web";
+    homepage = "https://thelounge.chat";
+    changelog = "https://github.com/thelounge/thelounge/releases/tag/v${finalAttrs.version}";
+    maintainers = with maintainers; [ winter ];
```
Comment (RaitoBezarius) [inline pkgs/applications/networking/irc/thelounge/default.nix:61]: you can add me too as a maintainer, no problem

**Review by RaitoBezarius — APPROVED** [review]
LGTM!

File: pkgs/applications/networking/irc/thelounge/default.nix [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  npmFlags = [ "--legacy-peer-deps" ];
+
+  # Takes way, way, way too long.
+  dontStrip = true;
+
+  passthru.tests = nixosTests.thelounge;
+
+  meta = with lib; {
+    description = "A glossy Matrix collaboration client for the web";
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply pkgs/applications/networking/irc/thelounge/default.nix:58]: yes yes it is :DDDDDDDDDDDDd

File: pkgs/applications/networking/irc/thelounge/default.nix (hunk 2 above)
Comment (winterqt) [reply pkgs/applications/networking/irc/thelounge/default.nix]: Sure, thanks.

**Comment by winterqt** [comment]
Waiting on OfBorg.

**Comment by winterqt** [comment]
Actually, let me try something else... I want to reduce this closure size.

**Comment by winterqt** [comment]
Before this push, it was just about 400 MiB, now it's 81 MiB.

**Review by RaitoBezarius — APPROVED** [review]
LGTM

File: pkgs/development/node-packages/aliases.nix [hunk 4]
```diff
@@ -44,4 +44,5 @@ mapAliases {
   "@nestjs/cli" = pkgs.nest-cli; # Added 2023-05-06
   manta = pkgs.node-manta; # Added 2023-05-06
   triton = pkgs.triton; # Added 2023-05-06
+  thelounge = pkgs.thelounge; # Added 2023-05-22
```
Comment (SuperSandro2000) [inline pkgs/development/node-packages/aliases.nix:47]:
```suggestion
  thelounge = pkgs.thelounge; # Added 2023-05-22
  triton = pkgs.triton; # Added 2023-05-06
```
nit

File: pkgs/applications/networking/irc/thelounge/default.nix [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    yarn install --offline --frozen-lockfile --ignore-platform --ignore-scripts --no-progress --non-interactive
+    patchShebangs node_modules
+
+    runHook postConfigure
+  '';
+
+  buildPhase = ''
+    runHook preBuild
+
+    NODE_ENV=production yarn build
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/irc/thelounge/default.nix:45]: we should also move the env to some hook but out of scope

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh [hunk 6]
```diff
@@ -27,7 +27,10 @@ npmInstallHook() {
     local -r nodeModulesPath="$packageOut/node_modules"
 
     if [ ! -d "$nodeModulesPath" ]; then
-        npm prune --omit dev --no-save $npmInstallFlags "${npmInstallFlagsArray[@]}" $npmFlags "${npmFlagsArray[@]}"
+        if [ -z "${dontNpmPrune-}" ]; then
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh:30]: This will be taken care in further PRs.

**Review by lilyinstarlight — APPROVED** [review]
This looks good to me too. I'll second Sandro's minor lexicographical ordering nit, but the `dontNpmPrune` install hook change looks fine enough for me

(I wish it were reasonable to separate prune/install into separate hooks entirely, but it really sorta has to happen in the middle there, so this is acceptable for now)

File: pkgs/applications/networking/irc/thelounge/default.nix (hunk 5 above)
Comment (winterqt) [reply pkgs/applications/networking/irc/thelounge/default.nix:45]: Agreed. However, this isn't the time for it due to the whole tooling shuffling/etc.

File: pkgs/development/node-packages/aliases.nix (hunk 4 above)
Comment (winterqt) [reply pkgs/development/node-packages/aliases.nix:47]: Thanks.

**Comment by winterqt** [comment]
Should be good to go now. Waiting on OfBorg.

File: pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (winterqt) [reply pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh:30]: Any reason why we shouldn't just do it here?

[the remark this answers is not in the document]
Comment (winterqt) [reply pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh:30]: Actually, we don't have any of the other `dont*` flags here, so let's not.

**Comment by winterqt** [comment]
I need more details, please -- what system are you running it on? This seemingly can't be reproduced in CI, or on Ryan's end.

**Comment by shyim** [comment]
I have the same aarch64-linux my config: https://github.com/shyim/nixcfg/blob/main/systems/shea/services/thelounge.nix

**Comment by Tom-Hubrecht** [comment]
> I need more details, please -- what system are you running it on? This seemingly can't be reproduced in CI, or on Ryan's end.

I am running thelounge on an x86_64 server, the config is at https://git.hubrecht.ovh/hubrecht/nixos/src/branch/master/machines/sd-02/thelounge.nix

The issue seems to be that old messages are no longer shown, but new messages still are (I am next to Ryan's guinea pig who confirmed the issue)

**Comment by RaitoBezarius** [comment]
> I need more details, please -- what system are you running it on? This seemingly can't be reproduced in CI, or on Ryan's end.

I confirm that my users doesn't see their messages, but the logs are on disk and the SQLite DB is intact.

**Comment by winterqt** [comment]
Let me look into this, please give me a bit.

**Comment by RaitoBezarius** [comment]
I do have this:
```
May 23 16:15:00 amadeus thelounge[108838]: 2023-05-23 14:15:00 [ERROR] Unable to load sqlite3 module. See https://github.com/mapbox/node-sqlite3/wiki/Binaries
May 23 16:15:00 amadeus thelounge[108838]: 2023-05-23 14:15:00 [INFO] The Lounge v4.4.0 (Node.js 18.16.0 on linux x64)
May 23 16:15:00 amadeus thelounge[108838]: 2023-05-23 14:15:00 [INFO] Configuration file: /var/lib/thelounge/config.js
May 23 16:15:01 amadeus thelounge[108838]: 2023-05-23 14:15:01 [INFO] Available at http://127.0.0.1:9000/ in private mode
```

in my systemd logs, FWIW.

@winterqt Thank you for looking into it (and I have no worries, I have backups and all.) :)

It's annoying that upstream didn't make the error fatal… :/

**Pull request merged**